x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java

```diff
@@ -0,0 +1,199 @@
+        } else {
+            this.type = Type.USER;
+        }
+        this.metadata = metadata;
+    }
+
+    public User getUser() {
+        return user;
+    }
+
+    public Authentication.RealmRef getRealm() {
+        return realm;
+    }
+
+    public Type getType() {
+        return type;
+    }
+
+    public Map<String, Object> getMetadata() {
+        return metadata;
+    }
+
+    /**
+     * Return a List of RoleReferences that represents role definitions associated to the subject.
+     * The final role of this subject should be the intersection of all role references in the list.
+     */
+    public List<RoleReference> getRoleReferences(@Nullable AnonymousUser anonymousUser) {
+        switch (type) {
+            case USER:
+                return buildRoleReferencesForUser(anonymousUser);
+            case API_KEY:
+                return buildRoleReferencesForApiKey();
+            case SERVICE_ACCOUNT:
+                return List.of(new RoleReference.ServiceAccountRoleReference(user.principal()));
+            default:
+                assert false : "unknown subject type: [" + type + "]";
+                throw new IllegalStateException("unknown subject type: [" + type + "]");
+        }
+    }
+
+    private List<RoleReference> buildRoleReferencesForUser(AnonymousUser anonymousUser) {
+        if (user.equals(anonymousUser)) {
+            return List.of(new RoleReference.NamedRoleReference(user.roles()));
+        }
+        final String[] allRoleNames;
+        if (anonymousUser == null || false == anonymousUser.enabled()) {
+            allRoleNames = user.roles();
+        } else {
+            // TODO: should we validate enable status and length of role names on instantiation time of anonymousUser?
+            if (anonymousUser.roles().length == 0) {
+                throw new IllegalStateException("anonymous is only enabled when the anonymous user has roles");
+            }
+            allRoleNames = Stream.concat(Arrays.stream(user.roles()), Arrays.stream(anonymousUser.roles())).toArray(String[]::new);
+        }
+        return List.of(new RoleReference.NamedRoleReference(allRoleNames));
+    }
+
+    private List<RoleReference> buildRoleReferencesForApiKey() {
+        if (false == version.onOrAfter(VERSION_API_KEY_ROLES_AS_BYTES)) {
+            return buildRolesReferenceForApiKeyBwc();
+        }
+        final String apiKeyId = (String) metadata.get(AuthenticationField.API_KEY_ID_KEY);
+        final BytesReference roleDescriptorsBytes = (BytesReference) metadata.get(API_KEY_ROLE_DESCRIPTORS_KEY);
+        final BytesReference limitedByRoleDescriptorsBytes = getLimitedByRoleDescriptorsBytes();
+        final RoleReference.ApiKeyRoleReference limitedByRoleReference = new RoleReference.ApiKeyRoleReference(
+            apiKeyId,
+            limitedByRoleDescriptorsBytes,
+            "apikey_limited_role"
+        );
+        if (roleDescriptorsBytes == null && limitedByRoleDescriptorsBytes == null) {
```

Comment:
Should this be moved up 1 statement, before we try to use `limitedByRoleDescriptorsBytes` ?

Reply:
Yes it should. Thanks for catching it.